Jellyfin.Api/Auth/CustomAuthenticationHandler.cs

```diff
@@ -64,5 +96,183 @@ protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+                || !Request.Headers.TryGetValue(HeaderNames.Authorization, out auth))
+            {
+                return null;
+            }
+
+            // Take first auth auth header.
+            var authHeaderValue = auth[0];
+            var parts = authHeaderValue.Split(new[] { ' ' }, 2);
+            if (parts.Length != 2)
+            {
+                return null;
+            }
+
+            var acceptedNames = new[] { "Jellyfin", "MediaBrowser", "Emby" };
+
+            // It has to be a digest request
+            if (!acceptedNames.Contains(parts[0], StringComparer.OrdinalIgnoreCase))
+            {
+                return null;
+            }
+
+            // Remove until the first space
+            authHeaderValue = parts[1];
+            parts = authHeaderValue.Split(',');
+
+            var result = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var item in parts)
+            {
+                var param = item.Trim().Split(new[] { '=' }, 2);
+
+                if (param.Length == 2)
+                {
+                    var value = NormalizeValue(param[1].Trim(new[] { '"' }));
+                    result.Add(param[0], value);
+                }
+            }
+
+            return result;
+        }
+
+        private AuthorizationInfo? GetTokenClaims(Dictionary<string, string>? authenticationClaims)
+        {
+            string? deviceId = null;
+            string? device = null;
+            string? client = null;
+            string? version = null;
+            string? token = null;
+            authenticationClaims?.TryGetValue("DeviceId", out deviceId);
+            authenticationClaims?.TryGetValue("Device", out device);
+            authenticationClaims?.TryGetValue("Client", out client);
+            authenticationClaims?.TryGetValue("Version", out version);
+            authenticationClaims?.TryGetValue("Token", out token);
+
+            var info = new AuthorizationInfo
+            {
+                Client = client,
+                Device = device,
+                DeviceId = deviceId,
+                Version = version,
+                Token = token
+            };
+
+            if (string.IsNullOrEmpty(token))
+            {
+                Request.Headers.TryGetValue("X-Jellyfin-Token", out var value);
+                token = value;
+            }
+
+            if (string.IsNullOrEmpty(token))
+            {
+                Request.Headers.TryGetValue("X-Emby-Token", out var value);
+                token = value;
+            }
+
+            if (string.IsNullOrEmpty(token))
+            {
+                Request.Headers.TryGetValue("X-MediaBrowser-Token", out var value);
+                token = value;
+            }
+
+            if (string.IsNullOrEmpty(token))
+            {
+                Request.Query.TryGetValue("api_key", out var value);
+                token = value;
+            }
```

Comment:
Maybe we should add an optional `ApiKey` parameter to the query in 10.6.0 and deprecate `api_key` in the next major release to fix the annoying underscore.

Reply:
Any chance you could also add a TODO here so we don't forget about this?